### DISCUSSION TITLE
Parse.User usage.

### DISCUSSION THREAD

**Discussion by alessdiimperio** [netreconlab/Parse-Swift#136] [General]
Hello! I've recently migrated from ParseObjC till ParseSwift only to find that it too was old and now migrated over to ParseSwiftOG. during migrating to ParseSwiftOG i had a few issues that would be nice to discuss.

1. change from mutating func to func -> Self... that was an annoying change and compiled fine but the code kept throwing errors all over my codebase.. Took me some time to realize that the new Self was returned and i had to refactor the code to adopt the new change.

var object = SomeParseObject()
try await object.save()
var installation = try await Installation.current()
installation.object = try object.toPointer() <- fail...

--------

seeing as it's value type and no longer mutating i had to refactor similar code to use the returned object rather the previous valueType:

var object = SomeParseObject()
let savedObject = try await object.save() 
var installation = try await Installation.current()
installation.object = try savedObject.toPointer()

----------------------------------------------
Would've been nice with a mention in the migration guide 😆 

Another issue that i'm currently having is
1. when changing password it seems that the User.current() gets signed out? this wasn't the case before... can't find any documentation on proper practices for handling login/logout.
Also the entire installation is deleted on logout, how to avoid?
save installation objectId -> Installation.become(objectId) is this how to do it if we want to keep same installation even after logout?

2. When i log a user out using try await User.logout() 
the logout works but if i try to log back in it throws an error that i cannot log in an objectId that differs from User.current().objectId... shouldn't logout remove the "current" user?

**Comment by alessdiimperio** [comment]
Alright i've got my code running and working as it was prior to the migration.

Things that would've been nice to have known prior to migrating would have been.

1. If you perform a .save() on any entity, the previous stored property is obsolete and no longer valid.
so always overwrite your stored properties with the return of the  try await object.save() to have the correct value.
previously the functions were mutating which meant the valueType was updated in scope, however the code still compiled just fine but would fail when trying to get objectId from the previous valueType prior to .save() is objects need to be saved to have objectId.
So code compiled fine but kept throwing errors.

2. If you log out, everything is deleted from keychain, including installation. which seems strange as installation is per app install and not per user login. but i can see how some people might store sensitive things and accidently forget to remove etc. so no biggie.

so before running User.logout() persist the objectId of the installation. and after logout rung Installation.become(installationId) to keep a consistent installation for app installation, just make sure no sensitive data is persisted on the installation.

3. Changing password changes the token and technically logs you out??. Would've expected this handled internally by the SDK.
so after changing the password and recieving a new token, i had to run User.login() again 🤔  is there something here that's missing? no documentation on this.

if any of my findings are incorrect please do inform me  cheers!

**Comment by cbaker6** [comment]
Happy to hear you were able to get your setup working! If you haven't already, I recommend you checkout the following:

- [Parse Swift Playgrounds](https://github.com/netreconlab/Parse-Swift/tree/main/ParseSwift.playground/Pages) provides runnable code examples and explanations about how the SDK works
- [ParseSwiftOG vs. Other Parse SDKs](https://github.com/netreconlab/Parse-Swift/discussions/72) provides feature comparisons against the other Pare SDKs and discusses how ParseSwiftOG is the most complete
- [ParseSwiftOG API Documentation](https://swiftpackageindex.com/netreconlab/Parse-Swift/5.8.2/documentation/parseswift)
- [Example ParseSwift Apps and Related Repos](https://github.com/netreconlab/Parse-Swift#example-apps-and-frameworks) showcases apps and libraries built with ParseSwift
- [README](https://github.com/netreconlab/Parse-Swift)
- [parse-community/ParseSwift closed issues](https://github.com/parse-community/Parse-Swift/issues?q=is%3Aissue+is%3Aclosed) has a ton questions/issues I've answered in the past
- [ParseSwift discord](https://community.parseplatform.org/c/client-sdks/parseswift-sdk/25) has a ton questions I've answered in the past 

The majority of the items you asked about are answered in the aforementioned links and the Playgrounds demonstrate how to properly use the SDK.

If you see fit, please star the repo and spread the word about ParseSwiftOG to your colleagues and on your social media platforms!

**Reply by alessdiimperio** [reply]
The ParseSwiftOG API documentation gave 404 yesterday but perhaps there is some old link on the forums i clicked on and assumed the docs where MIA.

Thanks for the links. i'll look into them!

**Reply by alessdiimperio** [reply]
Looked through the links and didnt see any examples for changing user password

Changing the password changes the parse session token which logs the user out.
If i try to do any other request using the parse SDK it fails.

after changing the user password and saving i need to do User.login(username, password).

**Reply by cbaker6** [reply]
@alessdiimperio thanks for reporting this problem, I've opened an issue in #145 and hope to address this shortly.

**Reply by cbaker6** [reply]
Fixed in #147
